File: Ageng_based_simulation/Force_based_model/misc.py
import numpy as np
def normalize(v):
    norm = np.linalg.norm(v)
    if norm == 0: 
       return v
    return v / norm

def apply_periodic_distance(agent_i_pos, agent_j_pos, env):
    """Calculate minimum separation between two agents with periodic boundaries."""
    dx = agent_j_pos[0] - agent_i_pos[0]
    dy = agent_j_pos[1] - agent_i_pos[1]

    # Apply periodic boundaries along the x-axis
    if abs(dx) > env.width / 2:
        dx = dx - np.sign(dx) * env.width

    # Apply periodic boundaries along the y-axis
    if abs(dy) > env.height / 2:
        dy = dy - np.sign(dy) * env.height

    return np.array([dx, dy])

def closest_distance_between_ellipses(agent1, agent2, env):
    # separation_unit_vector = normalize(apply_periodic_distance(agent1.position, agent2.position, env))
    separation_unit_vector = normalize(agent2.position - agent1.position)

    velocity_unit_vector1 = normalize(agent1.velocity)
    velocity_unit_vector2 = normalize(agent2.velocity)
    angle1 = np.arccos(np.dot(velocity_unit_vector1, separation_unit_vector))
    angle2 = np.arccos(np.dot(velocity_unit_vector2, separation_unit_vector))
    q1 = ((np.cos(angle1) ** 2) / (agent1.a ** 2)) + ((np.sin(angle1) ** 2) / (agent1.b ** 2))
    q2 = ((np.cos(angle2) ** 2) / (agent2.a ** 2)) + ((np.sin(angle2) ** 2) / (agent2.b ** 2))
   #  print(q1)
   #  print(q2)

    r1 = np.sqrt(1 / q1)
    r2 = np.sqrt(1 / q2)
    
    h = np.linalg.norm(apply_periodic_distance(agent1.position, agent2.position, env)) - r1 - r2
    
    return h

def calculate_angle(velocity):
    # Compute the angle in degrees from the velocity vector
    return np.degrees(np.arctan2(velocity[1], velocity[0]))


def calculate_shortest_distance_between_point_and_line(agent, line_points):
    point1 = np.array(line_points[0])
    point2 = np.array(line_points[1])
    agent_pos = agent.position
    line_vector = point2 - point1
    line_length_squared = np.dot(line_vector, line_vector)
    point_to_agent_vector = agent_pos - point1
    t = np.dot(point_to_agent_vector, line_vector) / line_length_squared
    t = max(0, min(1, t))
    nearest_point_on_line = point1 + t * line_vector
    d = nearest_point_on_line - agent_pos
    d_mag = np.linalg.norm(d)
    d_unit = d / d_mag
    return d_mag, normalize(line_vector), nearest_point_on_line, point1, point2
    
    
def det_numpy(A, B):
    # Stack vectors A and B as columns of a 2x2 matrix
    matrix = np.column_stack((A, B))
    # Compute the determinant of the 2x2 matrix
    return np.linalg.det(matrix)



def contact_distance_between_two_ellipse(agent1, agent2):
    k1 = normalize(agent1.velocity)
    b1 = agent1.b
    a1 = agent1.a
    eta = (a1 / b1) - 1
    T = (1 /b1) * (np.identity(2) + ((b1 / a1) - 1) * np.outer(k1, k1))
    Tinv = b1 * (np.identity(2) + eta * np.outer(k1, k1))
    
    
def calculate_closest_distance(agent1, agent2):
    """
    Calculate the closest distance between two ellipses in 2D space.

    Parameters:
        agent1, agent2: Two objects representing the ellipses, each having the following attributes:
            - a: Semi-major axis of the ellipse.
            - b: Semi-minor axis of the ellipse.
            - velocity: A vector representing the orientation of the ellipse (this gives the direction of the semi-major axis).
            - position: A vector representing the center of the ellipse.
    
    Returns:
        The closest distance between the centers of the two ellipses.
    """

    # Step 1: Extract ellipse parameters
    a1, b1 = agent1.a, agent1.b
    a2, b2 = agent2.a, agent2.b
    k1 = normalize(agent1.velocity)  # Orientation of ellipse 1
    k2 = normalize(agent2.velocity)  # Orientation of ellipse 2
    position_diff = agent2.position - agent1.position  # Vector joining the centers of the ellipses

    # Step 2: Define the transformation matrix to transform ellipse 1 into a unit circle
    def transformation_matrix(a, b, k):
        I = np.identity(2)
        k_k = np.outer(k, k)  # Outer product of the direction vector
        T = (1 / b) * (I + (b/a - 1) * k_k)  # Transformation matrix
        return T

    # Transform the first ellipse into a unit circle
    T1 = transformation_matrix(a1, b1, k1)

    # Step 3: Transform the second ellipse using the same transformation matrix
    # Ellipse matrix (representing the shape of the ellipse in quadratic form)
    def ellipse_matrix(a, b, k):
        I = np.identity(2)
        k_k = np.outer(k, k)
        return (1 / b**2) * (I + (b**2 / a**2 - 1) * k_k)

    A2 = ellipse_matrix(a2, b2, k2)  # Shape matrix for ellipse 2
    A2_transformed = np.dot(np.dot(np.linalg.inv(T1), A2), np.linalg.inv(T1))  # Transforming ellipse 2

    # Step 4: Calculate the quartic equation to solve for the distance in the transformed space
    # For simplicity, using the transformed semi-major and semi-minor axes
    eigenvalues, _ = np.linalg.eig(A2_transformed)  # Eigenvalues of the transformed matrix
    a2_prime = 1 / np.sqrt(np.min(eigenvalues))  # Transformed semi-major axis
    b2_prime = 1 / np.sqrt(np.max(eigenvalues))  # Transformed semi-minor axis
    
    # The distance between the unit circle and the transformed ellipse can be computed analytically:
    def transformed_distance(a_prime, b_prime):
        # Approximation of the closest distance in transformed space
        return 1 + a_prime  # Special case from the paper, for external tangency
    
    d_prime = transformed_distance(a2_prime, b2_prime)

    # Step 5: Apply inverse transformation to get the real-world distance
    # Calculate the angle between the major axis of ellipse 1 and the vector joining the centers
    angle_between = np.dot(k1, position_diff / np.linalg.norm(position_diff))
    eccentricity1 = np.sqrt(1 - (b1**2 / a1**2))  # Eccentricity of ellipse 1
    
    # Apply inverse transformation to find the real-world distance
    distance = d_prime / np.sqrt(1 - eccentricity1**2 * angle_between**2)
    
    return distance
    
    
# def calculate_closest_distance(agent1, agent2):
#     """
#     Calculate the closest distance between two ellipses with semi-major axes a1, a2,
#     semi-minor axes b1, b2, and orientation vectors k1, k2.

#     Parameters:
#         a1, b1: Semi-major and semi-minor axes of ellipse 1.
#         a2, b2: Semi-major and semi-minor axes of ellipse 2.
#         k1, k2: Orientation unit vectors for ellipses 1 and 2.

#     Returns:
#         The closest distance between the centers of the two ellipses.
#     """
#     a1, b1 = agent1.a, agent1.b
#     a2, b2 = agent2.a, agent2.b
#     k1, k2 = normalize(agent1.velocity), normalize(agent2.velocity)
    
#     # Step 1: Transformation matrix T to transform E1 into a unit circle
#     e1 = np.sqrt(1 - (b1**2 / a1**2))
#     e2 = np.sqrt(1 - (b2**2 / a2**2))
    
#     # Equation (4): T is a transformation matrix that scales the ellipse
#     def transformation_matrix(a, b, k):
#         I = np.identity(2)
#         k_k = np.outer(k, k)
#         T = (1 / b) * (I + (b/a - 1) * k_k)
#         return T
    
#     T1 = transformation_matrix(a1, b1, k1)
    
#     epsilon = 1e-10  # A small number to prevent division by zero

#     # Step 2: Transform the second ellipse
#     dot_product = np.dot(T1, k2)
#     norm_dot_product = np.linalg.norm(dot_product)

#     # Add epsilon to avoid division by zero
#     if norm_dot_product < epsilon:
#         norm_dot_product = epsilon

#     d_hat = dot_product / norm_dot_product
    
#     # Step 3: Apply transformed distances (Equation 36)
#     def distance_transformed(a2, b2, e1, d_hat):
#         return (1 / b1) * (1 / np.sqrt(1 - (e1**2) * (np.dot(d_hat, k1))**2))
    
#     d_transformed = distance_transformed(a2, b2, e1, d_hat)
    
#     # Inverse transformation to get the actual distance (Equation 35)
#     distance = d_transformed / np.linalg.norm(T1)
    
#     return distance

File: Ageng_based_simulation/Force_based_mode_C/misc.cpp
#include <iostream>
#include <vector>
#include <array>
#include <cmath>
#include "Agent.h"  // Include your Agent class header
#include "Environment.h" // Include your Environment class header

// Function to normalize a vector
std::array<double, 2> normalize(const std::array<double, 2>& v) {
    double norm = std::hypot(v[0], v[1]);
    if (norm == 0) {
        return v;  // Return the original vector if the norm is 0
    }
    return {v[0] / norm, v[1] / norm};
}

// Function to apply periodic distance between two agent positions
std::array<double, 2> apply_periodic_distance(const std::array<double, 2>& agent_i_pos, const std::array<double, 2>& agent_j_pos, const Environment& env) {
    std::array<double, 2> dx = {agent_j_pos[0] - agent_i_pos[0], agent_j_pos[1] - agent_i_pos[1]};

    // Apply periodic boundaries along the x-axis
    if (std::abs(dx[0]) > env.width / 2) {
        dx[0] -= std::copysign(env.width, dx[0]);
    }

    // Apply periodic boundaries along the y-axis
    if (std::abs(dx[1]) > env.height / 2) {
        dx[1] -= std::copysign(env.height, dx[1]);
    }

    return dx;
}

// Function to calculate the closest distance between two ellipses
double closest_distance_between_ellipses(const Agent& agent1, const Agent& agent2, const Environment& env) {
    auto separation_unit_vector = normalize(apply_periodic_distance(agent1.position, agent2.position, env));
    auto velocity_unit_vector1 = normalize(agent1.velocity);
    auto velocity_unit_vector2 = normalize(agent2.velocity);
    double angle1 = std::acos(dot(velocity_unit_vector1, separation_unit_vector));
    double angle2 = std::acos(dot(velocity_unit_vector2, separation_unit_vector));

    double q1 = (std::cos(angle1) * std::cos(angle1) / (agent1.a * agent1.a)) + (std::sin(angle1) * std::sin(angle1) / (agent1.b * agent1.b));
    double q2 = (std::cos(angle2) * std::cos(angle2) / (agent2.a * agent2.a)) + (std::sin(angle2) * std::sin(angle2) / (agent2.b * agent2.b));

    double r1 = std::sqrt(1 / q1);
    double r2 = std::sqrt(1 / q2);
    
    double h = std::hypot(agent1.position[0] - agent2.position[0], agent1.position[1] - agent2.position[1]) - r1 - r2;
    
    return h;
}

// Function to calculate the angle from a velocity vector
double calculate_angle(const std::array<double, 2>& velocity) {
    return std::degrees(std::atan2(velocity[1], velocity[0]));
}

// Function to calculate the shortest distance between a point and a line
std::tuple<double, std::array<double, 2>, std::array<double, 2>, std::array<double, 2>, std::array<double, 2>>
calculate_shortest_distance_between_point_and_line(const Agent& agent, const std::array<std::array<double, 2>, 2>& line_points) {
    std::array<double, 2> point1 = line_points[0];
    std::array<double, 2> point2 = line_points[1];
    std::array<double, 2> agent_pos = agent.position;
    std::array<double, 2> line_vector = {point2[0] - point1[0], point2[1] - point1[1]};
    double line_length_squared = dot(line_vector, line_vector);
    std::array<double, 2> point_to_agent_vector = {agent_pos[0] - point1[0], agent_pos[1] - point1[1]};
    double t = dot(point_to_agent_vector, line_vector) / line_length_squared;
    t = std::clamp(t, 0.0, 1.0);  // Clamp t between 0 and 1

    std::array<double, 2> nearest_point_on_line = {point1[0] + t * line_vector[0], point1[1] + t * line_vector[1]};
    std::array<double, 2> d = {nearest_point_on_line[0] - agent_pos[0], nearest_point_on_line[1] - agent_pos[1]};
    double d_mag = std::hypot(d[0], d[1]);
    std::array<double, 2> d_unit = normalize(d);

    return {d_mag, d_unit, nearest_point_on_line, point1, point2};
}

// Function to calculate the determinant of a 2x2 matrix formed by two vectors A and B
double det_numpy(const std::array<double, 2>& A, const std::array<double, 2>& B) {
    return A[0] * B[1] - A[1] * B[0];
}

// Function to calculate the contact distance between two ellipses (to be implemented)
double contact_distance_between_two_ellipses(const Agent& agent1, const Agent& agent2) {
    // Placeholder function; actual implementation needed
    std::array<double, 2> k1 = normalize(agent1.velocity);
    double b1 = agent1.b;
    double a1 = agent1.a;
    double eta = (a1 / b1) - 1;

    // Implement the necessary transformation and distance calculations
    // ...
}

// Function to calculate the closest distance between two agents
double calculate_closest_distance(const Agent& agent1, const Agent& agent2) {
    double a1 = agent1.a;
    double b1 = agent1.b;
    double a2 = agent2.a;
    double b2 = agent2.b;

    std::array<double, 2> k1 = normalize(agent1.velocity);
    std::array<double, 2> k2 = normalize(agent2.velocity);
    
    double e1 = std::sqrt(1 - (b1 * b1 / (a1 * a1)));
    double e2 = std::sqrt(1 - (b2 * b2 / (a2 * a2)));

    auto transformation_matrix = [](double a, double b, const std::array<double, 2>& k) {
        std::array<std::array<double, 2>, 2> T;
        double k_k = k[0] * k[0] + k[1] * k[1];
        T[0][0] = (1.0 / b) * (1.0 + (b / a - 1) * k_k);
        T[0][1] = 0; // Modify as needed
        T[1][0] = 0; // Modify as needed
        T[1][1] = (1.0 / b) * (1.0 + (b / a - 1) * k_k);
        return T;
    };

    std::array<std::array<double, 2>, 2> T1 = transformation_matrix(a1, b1, k1);
    
    // Step 2: Transform the second ellipse
    std::array<double, 2> d_hat = normalize({T1[0][0] * k2[0], T1[1][0] * k2[1]});
    
    // Step 3: Apply transformed distances
    auto distance_transformed = [](double a2, double b2, double e1, const std::array<double, 2>& d_hat) {
        return (1.0 / b1) * (1.0 / std::sqrt(1 - e1 * e1 * dot(d_hat, k1) * dot(d_hat, k1)));
    };

    double d_transformed = distance_transformed(a2, b2, e1, d_hat);
    
    // Inverse transformation to get the actual distance
    double distance = d_transformed / std::sqrt(dot(T1[0], T1[0]) + dot(T1[1], T1[1]));
    
    return distance;
}
